#include <iostream>
#include <boost/asio.hpp>
#include <memory>
#include <vector>
#include <windows.h>
#include <thread>
#include <unordered_map>
#include <mutex>

#include "protocol.pb.h"
#include "PacketDispatcher.h"

#pragma pack(push, 1)
struct PacketHeader {
    uint16_t size;
    uint16_t id;
};
#pragma pack(pop)

using boost::asio::ip::tcp;

// ===============================================
// 1. ì „ë°© ì„ ì–¸ ë° ì „ì—­ ë³€ìˆ˜ (ë””ìŠ¤íŒ¨ì²˜, ì„¸ì…˜ ê´€ë¦¬)
// ===============================================
class GameConnection;
class ClientSession;

PacketDispatcher<GameConnection> g_game_dispatcher;
PacketDispatcher<ClientSession> g_gateway_dispatcher;

std::shared_ptr<GameConnection> g_gameConnection; // S2S (GameServer ì—°ê²°ìš©)

std::unordered_map<std::string, std::shared_ptr<ClientSession>> g_clientMap; // ì ‘ì† í´ë¼ì´ì–¸íŠ¸ ì„¸ì…˜ ë§µ
std::mutex g_clientMutex;


// ====================================================
// 2. GameConnection (Gateway -> GameServer S2S ì ‘ì†ìš©)
// ====================================================
class GameConnection : public std::enable_shared_from_this<GameConnection> {
private:
    tcp::socket socket_;
    boost::asio::io_context& io_context_;
    PacketHeader header_;
    std::vector<char> payload_buf_;

public:
    GameConnection(boost::asio::io_context& io_context)
        : socket_(io_context), io_context_(io_context) {
    }

    void Connect(const std::string& ip, short port) {
        try {
            tcp::resolver resolver(io_context_);
            auto endpoints = resolver.resolve(ip, std::to_string(port));
            boost::asio::connect(socket_, endpoints);
            std::cout << "[Gateway] ğŸ•¹ï¸ GameServer(S2S) 9000ë²ˆ í¬íŠ¸ì— ì„±ê³µì ìœ¼ë¡œ ì—°ê²°ë˜ì—ˆìŠµë‹ˆë‹¤!\n";
            ReadHeader();
        }
        catch (std::exception&) {
            // ê°•ì œ ì¢…ë£Œ(Crash)ë¥¼ ë§‰ê³  ì¹œì ˆí•˜ê²Œ ì•Œë ¤ì¤ë‹ˆë‹¤.
            std::cerr << "ğŸš¨ [ì¹˜ëª…ì  ì—ëŸ¬] GameServer(9000 í¬íŠ¸)ê°€ êº¼ì ¸ ìˆìŠµë‹ˆë‹¤! GameServerë¥¼ ë¨¼ì € ì¼œì£¼ì„¸ìš”!\n";            
            exit(1); // ì•ˆì „í•˜ê²Œ í”„ë¡œê·¸ë¨ ì¢…ë£Œ
        }
    }

    void Send(uint16_t pktId, const google::protobuf::Message& msg) {
        std::string payload;
        msg.SerializeToString(&payload);
        PacketHeader header;
        header.size = static_cast<uint16_t>(sizeof(PacketHeader) + payload.size());
        header.id = pktId;

        auto send_buf = std::make_shared<std::vector<char>>(header.size);
        memcpy(send_buf->data(), &header, sizeof(PacketHeader));
        memcpy(send_buf->data() + sizeof(PacketHeader), payload.data(), payload.size());

        auto self(shared_from_this());
        boost::asio::async_write(socket_, boost::asio::buffer(*send_buf),
            [this, self, send_buf](boost::system::error_code ec, std::size_t) {
                if (ec) std::cerr << "[Gateway] GameServerë¡œ S2S íŒ¨í‚· ì „ì†¡ ì‹¤íŒ¨\n";
            });
    }

private:
    void ReadHeader() {
        auto self(shared_from_this());
        boost::asio::async_read(socket_, boost::asio::buffer(&header_, sizeof(PacketHeader)),
            [this, self](boost::system::error_code ec, std::size_t length) {
                if (!ec) {
                    if (header_.size < sizeof(PacketHeader) || header_.size > 4096) return;
                    uint16_t payload_size = static_cast<uint16_t>(header_.size - sizeof(PacketHeader));
                    if (payload_size == 0) {
                        // selfì˜ ë³µì‚¬ë³¸(session_ptr)ì„ ë§Œë“¤ì–´ ë„˜ê¹ë‹ˆë‹¤.
                        auto session_ptr = self;
                        g_game_dispatcher.Dispatch(session_ptr, header_.id, nullptr, 0);
                        ReadHeader();
                    }
                    else {
                        payload_buf_.resize(payload_size);
                        ReadPayload(payload_size);
                    }
                }
                else std::cerr << "[Gateway] ğŸš¨ GameServerì™€ì˜ ì—°ê²°ì´ ëŠì–´ì¡ŒìŠµë‹ˆë‹¤!\n";
            });
    }

    void ReadPayload(uint16_t payload_size) {
        auto self(shared_from_this());
        boost::asio::async_read(socket_, boost::asio::buffer(payload_buf_.data(), payload_size),
            [this, self, payload_size](boost::system::error_code ec, std::size_t length) {
                if (!ec) {
                    // ì—¬ê¸°ë„ ë§ˆì°¬ê°€ì§€ë¡œ ë³µì‚¬ë³¸ì„ ë§Œë“¤ì–´ ë„˜ê¹ë‹ˆë‹¤.
                    auto session_ptr = self;
                    g_game_dispatcher.Dispatch(session_ptr, header_.id, payload_buf_.data(), payload_size);
                    ReadHeader();
                }
            });
    }
};


// ================================================
// 3. ClientSession (DummyClient -> Gateway ì ‘ì†ìš©)
// ================================================
class ClientSession : public std::enable_shared_from_this<ClientSession> {
private:
    tcp::socket socket_;
    PacketHeader header_;
    std::vector<char> payload_buf_;
    std::string account_id_ = "";

public:
    ClientSession(tcp::socket socket) noexcept : socket_(std::move(socket)) {}
    void start() { ReadHeader(); }
    void SetAccountId(const std::string& id) { account_id_ = id; }
    const std::string& GetAccountId() const { return account_id_; }

    void Send(uint16_t pktId, const google::protobuf::Message& msg) {
        std::string payload;
        msg.SerializeToString(&payload);
        PacketHeader header;
        header.size = static_cast<uint16_t>(sizeof(PacketHeader) + payload.size());
        header.id = pktId;

        auto send_buf = std::make_shared<std::vector<char>>(header.size);
        memcpy(send_buf->data(), &header, sizeof(PacketHeader));
        memcpy(send_buf->data() + sizeof(PacketHeader), payload.data(), payload.size());

        auto self(shared_from_this());
        boost::asio::async_write(socket_, boost::asio::buffer(*send_buf),
            [this, self, send_buf](boost::system::error_code ec, std::size_t) {});
    }

    void OnDisconnected() {
        if (!account_id_.empty()) {
            // ë§µì—ì„œ ì§€ìš°ê¸° ì „ì— GameServerë¡œ "ì´ ìœ ì € ë‚˜ê°”ë‹¤"ê³  ì•Œë ¤ì¤ë‹ˆë‹¤.
            if (g_gameConnection) {
                Protocol::GatewayGameLeaveReq leave_req;
                leave_req.set_account_id(account_id_);
                g_gameConnection->Send(Protocol::PKT_GATEWAY_GAME_LEAVE_REQ, leave_req);
            }

            // Gatewayì˜ ì„¸ì…˜ ë§µì—ì„œ ì‚­ì œ
            std::lock_guard<std::mutex> lock(g_clientMutex);
            g_clientMap.erase(account_id_);
            std::cout << "[Gateway] ìœ ì € ì ‘ì† ì¢…ë£Œ ë° ë§µì—ì„œ ì‚­ì œë¨: " << account_id_ << "\n";
            account_id_ = "";
        }
    }

private:
    void ReadHeader() {
        auto self(shared_from_this());
        boost::asio::async_read(socket_, boost::asio::buffer(&header_, sizeof(PacketHeader)),
            [this, self](boost::system::error_code ec, std::size_t length) {
                if (!ec) {
                    if (header_.size < sizeof(PacketHeader) || header_.size > 4096) return;
                    uint16_t payload_size = static_cast<uint16_t>(header_.size - sizeof(PacketHeader));
                    if (payload_size == 0) {
                        auto session_ptr = self;
                        g_gateway_dispatcher.Dispatch(session_ptr, header_.id, nullptr, 0);
                        ReadHeader();
                    }
                    else {
                        payload_buf_.resize(payload_size);
                        ReadPayload(payload_size);
                    }
                }
                else OnDisconnected();
            });
    }

    void ReadPayload(uint16_t payload_size) {
        auto self(shared_from_this());
        boost::asio::async_read(socket_, boost::asio::buffer(payload_buf_.data(), payload_size),
            [this, self, payload_size](boost::system::error_code ec, std::size_t length) {
                if (!ec) {
                    auto session_ptr = self;
                    g_gateway_dispatcher.Dispatch(session_ptr, header_.id, payload_buf_.data(), payload_size);
                    ReadHeader();
                }
                else OnDisconnected();
            });
    }
};


// ==========================================
// 4. ì¸ê²Œì„ íŒ¨í‚· í•¸ë“¤ëŸ¬ (Gateway ì „ìš©)
// ==========================================
void Handle_GatewayConnectReq(std::shared_ptr<ClientSession>& session, char* payload, uint16_t payloadSize) {
    Protocol::GatewayConnectReq req;
    if (req.ParseFromArray(payload, payloadSize)) {
        session->SetAccountId(req.account_id());

        {
            std::lock_guard<std::mutex> lock(g_clientMutex);
            g_clientMap[req.account_id()] = session;
        }

        Protocol::GatewayConnectRes res;
        res.set_success(true);
        session->Send(Protocol::PKT_GATEWAY_CLIENT_CONNECT_RES, res);
        std::cout << "[Gateway] ìœ ì €(" << req.account_id() << ") ì¸ê²Œì„ ì…ì¥ ìŠ¹ì¸ ë° ë“±ë¡ ì™„ë£Œ!\n";
    }
}

void Handle_ChatReq(std::shared_ptr<ClientSession>& session, char* payload, uint16_t payloadSize) {
    Protocol::ChatReq req;
    if (req.ParseFromArray(payload, payloadSize)) {
        std::cout << "[Chat] " << session->GetAccountId() << " : " << req.msg() << "\n";

        Protocol::ChatRes res;
        res.set_account_id(session->GetAccountId());
        res.set_msg(req.msg());

        std::lock_guard<std::mutex> lock(g_clientMutex);
        for (auto& pair : g_clientMap) {
            auto client_session = pair.second;
            if (client_session) {
                client_session->Send(Protocol::PKT_GATEWAY_CLIENT_CHAT_RES, res);
            }
        }
    }
}

// [í´ë¼ì´ì–¸íŠ¸ -> ê²Œì´íŠ¸ì›¨ì´] ìœ ì €ê°€ ì›€ì§ì˜€ì„ ë•Œ
void Handle_MoveReq(std::shared_ptr<ClientSession>& session, char* payload, uint16_t payloadSize) {
    Protocol::MoveReq req;
    if (req.ParseFromArray(payload, payloadSize)) {
        // ë‚´ IDë¥¼ ë¶™ì—¬ì„œ GameServerë¡œ S2S í† ìŠ¤ (ë¼ìš°íŒ…)
        if (g_gameConnection) {
            Protocol::GatewayGameMoveReq s2s_req;
            s2s_req.set_account_id(session->GetAccountId());
            s2s_req.set_x(req.x());
            s2s_req.set_y(req.y());
            s2s_req.set_z(req.z());
            s2s_req.set_yaw(req.yaw());

            g_gameConnection->Send(Protocol::PKT_GATEWAY_GAME_MOVE_REQ, s2s_req);
        }
    }
}

// [ê²Œì„ì„œë²„ -> ê²Œì´íŠ¸ì›¨ì´] S2S ì´ë™ ì§€ì‹œ íŒ¨í‚· ìˆ˜ì‹ 
void Handle_MoveRes_FromGame(std::shared_ptr<GameConnection>& conn, char* payload, uint16_t payloadSize) {
    Protocol::GameGatewayMoveRes s2s_res;
    if (s2s_res.ParseFromArray(payload, payloadSize)) {

        // í´ë¼ì´ì–¸íŠ¸ë“¤ì´ ë°›ì„ ì‹¤ì œ MoveRes íŒ¨í‚· ì„¸íŒ…
        Protocol::MoveRes client_res;
        client_res.set_account_id(s2s_res.account_id());
        client_res.set_x(s2s_res.x());
        client_res.set_y(s2s_res.y());
        client_res.set_z(s2s_res.z());
        client_res.set_yaw(s2s_res.yaw());

        std::lock_guard<std::mutex> lock(g_clientMutex);

        // â˜… í•µì‹¬: ì „ì²´ ë§µ(g_clientMap)ì„ ë¬´ì¡°ê±´ ë„ëŠ” ê²ƒì´ ì•„ë‹ˆë¼, 
        // GameServerê°€ ê³„ì‚°í•´ì„œ ì•Œë ¤ì¤€ íƒ€ê²Ÿ ëª…ë‹¨(AOI)ë§Œ ì™ì™ ë½‘ì•„ì„œ ì „ì†¡í•©ë‹ˆë‹¤!
        for (const std::string& target_id : s2s_res.target_account_ids()) {
            auto it = g_clientMap.find(target_id);
            if (it != g_clientMap.end()) {
                auto client_session = it->second;
                if (client_session) {
                    client_session->Send(Protocol::PKT_GATEWAY_CLIENT_MOVE_RES, client_res);
                }
            }
        }
    }
}

// ==========================================
// [ì¶”ê°€] GameServer -> Gateway -> Client ì „íˆ¬(í”¼ê²©) ë¸Œë¡œë“œìºìŠ¤íŠ¸ ë¦´ë ˆì´
// ==========================================
void Handle_GameGatewayAttackRes(std::shared_ptr<GameConnection>& session, char* payload, uint16_t size) {
    Protocol::GameGatewayAttackRes s2s_res;
    if (s2s_res.ParseFromArray(payload, size)) {

        // 1. í´ë¼ì´ì–¸íŠ¸ê°€ ì½ì„ ìˆ˜ ìˆëŠ” AttackRes íŒ¨í‚·ìœ¼ë¡œ ë³€í™˜ (í¬ì¥ì§€ êµì²´)
        Protocol::AttackRes client_res;
        client_res.set_attacker_uid(s2s_res.attacker_uid());
        client_res.set_target_account_id(s2s_res.target_account_id());
        client_res.set_damage(s2s_res.damage());
        client_res.set_target_remain_hp(s2s_res.target_remain_hp());

        // 2. ì´í™íŠ¸/ë°ë¯¸ì§€ë¥¼ ë´ì•¼ í•˜ëŠ” ì£¼ë³€ ìœ ì €(AOI)ë“¤ì—ê²Œ ê°ê° ì „ì†¡
        std::lock_guard<std::mutex> lock(g_clientMutex);
        for (int i = 0; i < s2s_res.target_account_ids_size(); ++i) {
            const std::string& target_id = s2s_res.target_account_ids(i);

            // í˜„ì¬ ê²Œì´íŠ¸ì›¨ì´ ì„¸ì…˜ ë§µì— ì ‘ì†í•´ ìˆëŠ” ìœ ì €ë¼ë©´ íŒ¨í‚· ì˜ê¸°!
            auto it = g_clientMap.find(target_id);
            if (it != g_clientMap.end() && it->second) {
                // PKT_GATEWAY_CLIENT_ATTACK_RES (27ë²ˆ) ìœ¼ë¡œ í´ë¼ì´ì–¸íŠ¸ì—ê²Œ ì „ì†¡
                it->second->Send(Protocol::PKT_GATEWAY_CLIENT_ATTACK_RES, client_res);
            }
        }
    }
}

// ==========================================
// 5. GatewayServer ìˆ˜ì‹  ëŒ€ê¸°ì—´ ë° Main
// ==========================================
class GatewayServer {
    tcp::acceptor acceptor_;
public:
    GatewayServer(boost::asio::io_context& io_context, short port)
        : acceptor_(io_context, tcp::endpoint(tcp::v4(), port)) {
        do_accept();
    }
private:
    void do_accept() {
        acceptor_.async_accept([this](boost::system::error_code ec, tcp::socket socket) {
            if (!ec) std::make_shared<ClientSession>(std::move(socket))->start();
            do_accept();
            });
    }
};

int main() {
    SetConsoleOutputCP(CP_UTF8);
    // ==========================================
    // [í´ë¼ì´ì–¸íŠ¸ -> ê²Œì´íŠ¸ì›¨ì´] íŒ¨í‚· í•¸ë“¤ëŸ¬ ë“±ë¡
    // ==========================================
    g_gateway_dispatcher.RegisterHandler(Protocol::PKT_CLIENT_GATEWAY_CONNECT_REQ, Handle_GatewayConnectReq);
    g_gateway_dispatcher.RegisterHandler(Protocol::PKT_CLIENT_GATEWAY_CHAT_REQ, Handle_ChatReq);
    g_gateway_dispatcher.RegisterHandler(Protocol::PKT_CLIENT_GATEWAY_MOVE_REQ, Handle_MoveReq);

    // ==============================================
    // [ê²Œì„ì„œë²„ -> ê²Œì´íŠ¸ì›¨ì´(S2S)] íŒ¨í‚· í•¸ë“¤ëŸ¬ ë“±ë¡
    // ==============================================
    // ì£¼ì˜: PKT_MOVE_RES(25)ê°€ ì•„ë‹ˆë¼ PKT_S2S_MOVE_RES(1025)ë¥¼ ë“±ë¡í•´ì•¼ í•©ë‹ˆë‹¤.
    g_game_dispatcher.RegisterHandler(Protocol::PKT_GAME_GATEWAY_MOVE_RES, Handle_MoveRes_FromGame);
    g_game_dispatcher.RegisterHandler(Protocol::PKT_GAME_GATEWAY_ATTACK_RES, Handle_GameGatewayAttackRes);

    try {
        boost::asio::io_context io_context;

        // â˜… 1. ìœ ì €ë¥¼ ë°›ê¸° ì „ì— GameServer(9000ë²ˆ)ë¡œ ë¨¼ì € S2S ì ‘ì† ì‹œë„
        g_gameConnection = std::make_shared<GameConnection>(std::ref(io_context));
        g_gameConnection->Connect("127.0.0.1", 9000);

        // â˜… 2. í´ë¼ì´ì–¸íŠ¸ ëŒ€ê¸°ì—´(8888ë²ˆ) ì˜¤í”ˆ
        GatewayServer server(io_context, 8888);
        std::cout << "[GatewayServer] ê²Œì„ ê²Œì´íŠ¸ì›¨ì´ ì„œë²„ ê°€ë™ ì‹œì‘ (Port: 8888) Created by Jeong Shin Young\n";
        std::cout << "=================================================\n";

        unsigned int thread_count = std::thread::hardware_concurrency();
        if (thread_count == 0) thread_count = 4;
        std::vector<std::thread> threads;
        for (unsigned int i = 0; i < thread_count; ++i) {
            threads.emplace_back([&io_context]() { io_context.run(); });
        }
        for (auto& t : threads) { if (t.joinable()) t.join(); }
    }
    catch (std::exception& e) { std::cerr << "[Error] " << e.what() << "\n"; }
    return 0;
}